src/core/inductive.ml

```diff
@@ -0,0 +1,195 @@
+              (x, Bindlib.subst b (Vari x))
+          in
+          let b = case ((Bindlib.box_var x)::xs) b in
+          begin
+            match unfold a with
+            | Symb(s) ->
+                let b =
+                  if s == sind then _Impl (prf_of_p (Bindlib.box_var x)) b
+                  else b
+                in
+              _Prod (Bindlib.box a) (Bindlib.bind_var x b)
+            | _ -> fatal pos "The type of %a is not supported"
+                     pp_symbol scons
+          end
+      | _ ->
+          fatal pos "The type of %a is not supported"
+            pp_symbol scons
+    in
+    case [] !(scons.sym_type)
+  in
+
+  let t =
+    let x = Bindlib.new_var mkfree "x" in
+    _Prod ind (Bindlib.bind_var x (prf_of_p (_Vari x)))
+  in
+  (* let add_case t scons = _Impl (case_of scons) t in
+     let t = List.fold_left add_case t scons_list   in *)
+  let rec add_case l r = match l with
+    | []   -> r
+    | t::q -> let t = case_of t in
+              _Impl t (add_case q r)
+  in
+  let t = add_case scons_list t in
+  let t = _Prod (_Impl ind prop) (Bindlib.bind_var p t) in
+  Bindlib.unbox t
+
+let ind_rule : Sig_state.t -> popt -> string -> string -> term -> sym list
+               -> p_rule list =
+  fun _ pos type_name ind_prop_name ind_prop_type cons_list ->
+  (* Find the induction principale *)
+  (* let i =
+       try SymMap.find sind !(ss.signature.sign_ind)
+       with Not_found -> assert false
+     in*)
+    (* Create the common head of the rules *)
+  let arg : sym list -> qident -> p_term = fun l ind_prop ->
+    let p = Pos.make pos "p" in
+    let p_iden = Pos.make pos (P_Iden(ind_prop, true)) in
+    let p_patt = Pos.make pos (P_Patt(Some p, [||]))   in
+    let head = P_Appl(p_iden, p_patt)                  in
+    let rec aux : sym list -> p_term -> p_term = fun l acc ->
+      match l with
+      | []   -> acc
+      | t::q ->
+          let t = Pos.make pos ("p" ^ t.sym_name)           in
+          let p_patt = Pos.make pos (P_Patt(Some t, [||])) in
+          aux q (Pos.make pos (P_Appl(acc, p_patt)))
+    in
+    aux l (Pos.make pos head)
+  in
+  let common_head = arg cons_list (Pos.make pos ([], ind_prop_name)) in
+  if !log_enabled then
+    log_induc_rule "The common head of the rules [%a]"
+      Pretty.pp_p_term common_head;
+  (* Build the whole of the rules *)
+  let e : term -> sym list -> p_rule list = fun _ l ->
+    let rec aux : sym list -> p_rule list -> p_rule list = fun l acc ->
+      match l with
+      | []   -> acc
+      | t::q ->
+          begin
+          let pt       = Pos.make pos ("p" ^ t.sym_name)       in (* RHS *)
+          let p_patt   = Pos.make pos (P_Patt(Some pt, [||]))  in
+          let qident_t = Pos.make pos ([], t.sym_name)         in (* LHS *)
+          let t_ident  = Pos.make pos (P_Iden(qident_t, true)) in
+          let tmp      = aux q acc                             in
+          let rec truc arg_list t hyp_rec_list = match unfold t with
```

Comment:
Bad name...

Reply:
If you use the same iteration scheme as in `case_of`, it would be good to try to factorize them and define a more general case_of function.